### ISSUE TITLE
Patroni should failover when the disk on the primary is full

### ISSUE THREAD

**Issue by JoseFinottoGitLab** [zalando/patroni#1732]
**Describe the bug**
We had a scenario where the primary got the disk full in staging, and I think will be a desired behaviour that a failover will happen.

**To Reproduce**
Due to a replication slot that was saving data, the disk got full on the primary. and it restarted the node, without promotion for further details : https://gitlab.com/gitlab-com/gl-infra/infrastructure/-/issues/11210

 
**Expected behavior**

It is expected that patroni would failover in case of the data disk get full.

**Screenshots**

**Environment**
- Patroni version: 1.6.0
- PostgreSQL version: 11.7
- DCS (and its version): Consul 

**Patroni configuration file**
```
---
scope: pg11-ha-cluster
name: patroni-07-db-gstg.c.gitlab-staging-1.internal
restapi:
  listen: 0.0.0.0:8009
  connect_address: 10.224.29.107:8009
consul:
  host: 127.0.0.1:8500
  checks: []
bootstrap:
  dcs:
    ttl: 90
    loop_wait: 10
    retry_timeout: 40
    maximum_lag_on_failover: 1048576
    postgresql:
      use_pg_rewind: true
      use_slots: true
      parameters:
        wal_level: logical
        hot_standby: 'on'
        wal_keep_segments: 512
        max_wal_senders: 32
        max_replication_slots: 32
        checkpoint_timeout: 10min
        max_connections: 500
        max_locks_per_transaction: 128
        max_wal_size: 16GB
  initdb:
  - encoding: UTF8
  - locale: C.UTF-8
  pg_hba:
  - hostssl all gitlab-superuser all md5
  - hostssl replication gitlab-replicator 127.0.0.1/32 md5
  - hostssl replication gitlab-replicator all md5
  - host postgres gitlab-superuser 10.0.0.0/8 md5
  - host all gitlab-superuser 10.0.0.0/8 md5
  - host all gitlab-superuser 127.0.0.1/32 md5
  - host replication gitlab-replicator 127.0.0.1/32 md5
  users:

tags: {}
postgresql:
  data_dir: "/var/opt/gitlab/postgresql/data11"
  config_dir: "/var/opt/gitlab/postgresql"
  bin_dir: "/usr/lib/postgresql/11/bin"
  listen: 0.0.0.0:5432
  parameters:
    port: 5432
    ssl: 'on'
    ssl_ciphers: HIGH:MEDIUM:+3DES:!aNULL:!SSLv3:!TLSv1
    log_destination: csvlog
    archive_command: "/opt/wal-g/bin/archive-wal.sh %p"
    archive_mode: 'on'
    archive_timeout: 1min
    autovacuum_analyze_scale_factor: '0.005'
    autovacuum_max_workers: '10'
    autovacuum_vacuum_cost_delay: 5ms
    autovacuum_vacuum_cost_limit: 6000
    autovacuum_vacuum_scale_factor: '0.005'
    checkpoint_completion_target: '0.9'
    deadlock_timeout: 5s
    default_statistics_target: '1000'
    default_text_search_config: pg_catalog.english
    effective_cache_size: 338688MB
    effective_io_concurrency: 100
    hba_file: "/var/opt/gitlab/postgresql/data11/pg_hba.conf"
    hot_standby_feedback: 'on'
    idle_in_transaction_session_timeout: 30s
    lc_messages: C
    log_autovacuum_min_duration: '0'
    log_checkpoints: 'on'
    log_line_prefix: "%m [%p, %x]: [%l-1] user=%u,db=%d,app=%a,client=%h "
    log_connections: 'on'
    log_disconnections: 'on'
    log_statement: ddl
    log_lock_waits: 'on'
    log_min_duration_statement: 1000
    log_temp_files: 0
    logging_collector: 'on'
    log_directory: "/var/log/gitlab/postgresql"
    log_filename: postgresql.log
    log_file_mode: '0640'
    log_rotation_age: 0
    log_rotation_size: 0
    maintenance_work_mem: 2048MB
    random_page_cost: 1.5
    shared_buffers: 7538MB
    shared_preload_libraries: pg_stat_statements
    ssl_ca_file: ''
    statement_timeout: '15000'
    tcp_keepalives_count: 0
    tcp_keepalives_idle: 0
    tcp_keepalives_interval: 0
    track_activity_query_size: '1024'
    track_io_timing: 'on'
    unix_socket_directories: "/var/opt/gitlab/postgresql"
    wal_compression: 'on'
    wal_log_hints: 'on'
    work_mem: 100MB
```

**patronictl show-config**
```
loop_wait: 10
maximum_lag_on_failover: 1048576
postgresql:
  parameters:
    checkpoint_timeout: 10min
    hot_standby: 'on'
    max_connections: 500
    max_locks_per_transaction: 128
    max_replication_slots: 32
    max_wal_senders: 32
    max_wal_size: 16GB
    wal_keep_segments: 512
    wal_level: logical
  use_pg_rewind: true
  use_slots: true
retry_timeout: 40
ttl: 90
```

**Additional context**
Issue with more details: https://gitlab.com/gitlab-com/gl-infra/infrastructure/-/issues/11210

**Comment by le-duane** [comment]
@JoseFinottoGitLab this issue has come up a few times before https://github.com/zalando/patroni/issues?q=is%3Aissue+disk+full

**Comment by bradnicholson** [comment]
Patroni will failover if the primary runs out of disk space you have a secondary that is up to date enough to failover to, so there is more to this picture than what is reported.

I wonder if this is related to the master_start_timeout bug that was just fixed in 2.0.1?

**Comment by CyberDem0n** [comment]
The master_start_timeout fixes were related to crash-recovery in a single-user mode.

Besides that, there is nothing Patroni could do, but https://github.com/zalando/patroni/pull/1839 will help to address the issue by implementing custom monitoring scripts on the side of the end-user.
